### PR TITLE
add script to compute precomputed curves, test with precomputed

### PR DESCRIPTION
**Pull request by mvayngrib**
current master is forked from elliptic@6.4.1, renamed to @exodus/elliptic

the goal is to be able to sacrifice bundle size for performance by electing to use precomputed curves from the root project, e.g.:

```js
const elliptic = require('elliptic');
elliptic.usePrecomputed({
  // require only the ones you need
  p256: require('elliptic/lib/elliptic/precomputed/p256'),
});
```

also, the validation in the PresetCurve is expensive and unnecessary to do at runtime for pre-defined curves

### CODE REVIEW

File: lib/elliptic/curves.js [hunk 1]
```diff
@@ -18,8 +18,10 @@ function PresetCurve(options) {
   this.n = this.curve.n;
   this.hash = options.hash;
 
-  assert(this.g.validate(), 'Invalid curve');
-  assert(this.g.mul(this.n).isInfinity(), 'Invalid curve, G*N != O');
+  if (!skipValidation) {
+    assert(this.g.validate(), 'Invalid curve');
+    assert(this.g.mul(this.n).isInfinity(), 'Invalid curve, G*N != O');
```
Comment (mvayngrib) [inline lib/elliptic/curves.js]: this validation is expensive, and is not necessary for the curves defined in this file

File: lib/elliptic/use-precomputed.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+var curves = require('./curves');
+
+function usePrecomputed(precomputed) {
```
Comment (kklash) [inline lib/elliptic/use-precomputed.js:3]: Since the precomputed curves are stored inside the package, why do we make the caller specify them with an awkward internal require? Why not have `precomputed` be a string or some other constant which identifies the curve, and the package handles finding the curve's JSON file.

**Review by kklash — CHANGES_REQUESTED** [review]

File: lib/elliptic/curves.js [hunk 3]
```diff
@@ -7,7 +7,7 @@ var elliptic = require('../elliptic');
 
 var assert = elliptic.utils.assert;
 
-function PresetCurve(options) {
+function PresetCurve(options, skipValidation) {
```
Comment (kklash) [inline lib/elliptic/curves.js]: I feel like this should be part of the `options` object.

File: lib/elliptic/use-precomputed.js (hunk 2 above)
Comment (mvayngrib) [reply lib/elliptic/use-precomputed.js:3]: it's so the caller can decide what gets bundled (e.g. we might only want to require() the p256 precomputed json, and wouldn't want the others to end up in the final JS bundle). I agree that it's awkward, and i'm open to a simpler way if you know one :)

File: lib/elliptic/curves.js (hunk 3 above)
Comment (mvayngrib) [reply lib/elliptic/curves.js]: agree, i separated it so it wouldn't get passed through to elliptic.curve.short/elliptic.curve.mont/elliptic.curve.edwards, but i can merge it into options if u prefer

Comment (mvayngrib) [reply lib/elliptic/curves.js]: updated

**Review by ChALkeR — COMMENTED** [review]
Could there be a test that all precomputed curves match with non-precomputed ones?
Similar to `precompute.js`, but for verification only and automated during tests.

File: lib/elliptic/use-precomputed.js (hunk 2 above)
Comment (kklash) [reply lib/elliptic/use-precomputed.js:3]: Ah thank you, that makes more sense. 

What if the actual precomputed curve files could be packaged separately? This package would have the logic of how to use the precomputed curves, while a separate NPM package holds the curve JSON files themselves, using `elliptic` as a devDependency to generate them.

I don't think we need to do that right now though, just a future todo.

**Comment by kklash** [comment]
Agreed with @ChALkeR, that seems like a must-have opportunity for a unit test

**Comment by mvayngrib** [comment]
@kklash @ChALkeR see `scripts` for `unit:precomputed`. It runs the same tests with precomputed curves. You can also delete the contents of the `precomputed/` dir and re-gen them with `./scripts/precompute.js` and see that the git diff is empty

**Comment by mvayngrib** [comment]
> Could there be a test that all precomputed curves match with non-precomputed ones?

@ChALkeR in case i misunderstood, in which way did you mean for them to match?

File: lib/elliptic/use-precomputed.js (hunk 2 above)
Comment (mvayngrib) [reply lib/elliptic/use-precomputed.js:3]: yes, we could put them in a separate package, say `@exodus/precomputed-curves`, but the tests would be a bit harder to do than directly inside `elliptic`

we could do a mix of the two: gen precomputed curves and test them here, but re-export a cleaner api for `usePrecomputed` from that separate package

**Comment by kklash** [comment]
> in case i misunderstood, in which way did you mean for them to match?

I imagine he means that if you run regular old `elliptic` with no precomputed curves, and generate a `p256` curve, for example, that curve's JSON object should exactly match the precomputed `p256.json` object. I assume you could check using some kind of deep equality comparison function.

**Comment by mvayngrib** [comment]
@kklash ok, i'll see if i can add some kind of deep equality check to this test tomorrow: https://github.com/ExodusMovement/elliptic/pull/1/files#diff-eb6a8b288bf31a87c6e73e3e18055362R32

**Comment by mvayngrib** [comment]
@ChALkeR did i understand you correctly with this test? https://github.com/ExodusMovement/elliptic/pull/1/commits/83d7b7c4eded8799016fcdfbb00492eb55933584

File: test/use-precomputed-test.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      .filter(function (file) {
+        return path.extname(file) === '.json'
+      })
+      .forEach(function (file) {
+        var name = path.parse(file).name;
+        console.log('comparing', name)
+        var precomputed = require(path.join(dir, file));
+        var curve = elliptic.ec(name);
+        var computed = JSON.parse(JSON.stringify(curve.g.toJSON()));
+        assert.deepEqual(
```
Comment (ChALkeR) [inline test/use-precomputed-test.js]:
```suggestion
        assert.deepStrictEqual(
```

Comment (mvayngrib) [reply test/use-precomputed-test.js]: merged

**Comment by mvayngrib** [comment]
@ChALkeR @kklash is anything still missing from the PR?

**Comment by mvayngrib** [comment]
increased timeout for test

can i remove support for node < 4 and the coveralls test so we don't waste time on them? Or do we want to setup coveralls here?

the failing build for reference: https://travis-ci.com/ExodusMovement/elliptic/builds/121961197

File: lib/elliptic/curves.js (hunk 1 above)
Comment (ChALkeR) [reply lib/elliptic/curves.js]: @mvayngrib Are those validated in tests? I don't see those checks in that exact form.

**Review by ChALkeR — COMMENTED** [review]
The checks removed in `options.skipValidation` block should probably be extracted into a testcase, to make sure that all hardcoded curves do not fail that check. I don't think those are present in tests?
I.e.
```js
assert(this.g.validate(), 'Invalid curve');
assert(this.g.mul(this.n).isInfinity(), 'Invalid curve, G*N != O');
```
---

As for older Node.js versions (0.10, 0.12) support -- those fail due to new syntax usage,
so we should either use older syntax to support those, or just drop those from support and test matrix.
I would suggest removing those, those reached EOL a long time ago.

**Comment by mvayngrib** [comment]
> The checks removed in options.skipValidation block should probably be extracted into a testcase

good catch! pushed a test

**Review by kklash — APPROVED** [review]
utACK

File: README.md [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# Elliptic [![Build Status](https://secure.travis-ci.org/indutny/elliptic.png)](http://travis-ci.org/indutny/elliptic) [![Coverage Status](https://coveralls.io/repos/indutny/elliptic/badge.svg?branch=master&service=github)](https://coveralls.io/github/indutny/elliptic?branch=master)
+# Elliptic [![Build Status](https://secure.travis-ci.org/ExodusMovement/elliptic.png)](http://travis-ci.org/ExodusMovement/elliptic) [![Coverage Status](https://coveralls.io/repos/ExodusMovement/elliptic/badge.svg?branch=master&service=github)](https://coveralls.io/githubExodusMovement/elliptic?branch=master)
```
Comment (ChALkeR) [inline README.md]: Mistype: `githubExodusMovement` -> `github/ExodusMovement`

**Review by ChALkeR — APPROVED** [review]
utACK.

File: README.md (hunk 5 above)
Comment (mvayngrib) [reply README.md]: fixed

**Review by gutenye — APPROVED** [review]
utACK

**Review by Sekhmet — APPROVED** [review]
tACK

Tested:
- generating curves using the script.

**Comment by unidwell** [comment]
TLDR: Using precomputations **does not work** for multiple reasons. But, ecliptic library works as before only without precomputations. Will review again when fixed.

**Simple proof**

Default case, precomputed is calculated on the fly. It displays **true**.

```js
var ec = elliptic.ec('p256');
console.log('isPrecomputed:', ec.g._hasDoubles(new BN(1)));
```

`usePrecomputed` does not work; it displays **false**.

```js
elliptic.usePrecomputed({
  p256: require('../lib/elliptic/precomputed/p256');
});

var ec = elliptic.ec('p256');
console.log('isPrecomputed:', ec.g._hasDoubles(new BN(1)));
```

**Explanation**

It does not work, because **toJSON** is not matched somewhere with **fromJSON**. We have:
```
curves.forEach(({ name, curve }) => {
  const precomputed = curve.g.toJSON()
  fs.writeFileSync(getCurvePath(name), prettify(precomputed))
})
```

The EC point's **toJSON** creates a transformed JSON. We were unlucky the following did not break when curves are being loaded:

```js
names.forEach(function(name) {
  if (!curves[name].g.precomputed) {
    curves[name].g.precomputed = precomputed[name];
  }
});
```

**Possible solution**

We patch `elliptic/curve/index.js` EC constructor like this (new option **fromJSON**):

```js
function EC(options, fromJSON) {
  if (!(this instanceof EC))
    return new EC(options, fromJSON);

  // Shortcut `elliptic.ec(curve-name)`
  if (typeof options === 'string') {
    assert(elliptic.curves.hasOwnProperty(options), 'Unknown curve ' + options);

    options = elliptic.curves[options];
  }

  // Shortcut for `elliptic.ec(elliptic.curves.curveName)`
  if (options instanceof elliptic.curves.PresetCurve)
    options = { curve: options };

  this.curve = options.curve.curve;
  this.n = this.curve.n;
  this.nh = this.n.ushrn(1);
  this.g = this.curve.g;

  // Point on curve
  this.g = options.curve.g;
  if (fromJSON) {
    this.g = options.curve.curve.pointFromJSON(fromJSON)
  }
  else {
    this.g.precompute(options.curve.n.bitLength() + 1);
  }

  // Hash for function for DRBG
  this.hash = options.hash || options.curve.hash;
}
```

This will make the following example work correctly.  It displays **true**:

```js
var ec = elliptic.ec('p256', require('../lib/elliptic/precomputed/p256'));
console.log('isPrecomputed:', ec.g._hasDoubles(new BN(1)));
```

If we need complete compatibility with existing libs, it should be possible to monkey patch EC, so that `elliptic.ec('p256')` returns `elliptic.ec('p256', require('../lib/elliptic/precomputed/p256'))`

**Unanswered questions**

- Currently only operations with EC point **g** are accelerated (the original library and our library). Not sure if that is necessary. A `console.log` in `_hasDoubles` can show, that for instance, multiplication could benefit two times, but  **precomputed** is successfully used only once. Precomputations are probably still WIP in original lib.

File: test/use-precomputed-test.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    assert(g.precomputed == null, 'g.precomputed stars out null');
+
+    elliptic.usePrecomputed({
+      p256: require('../lib/elliptic/precomputed/p256')
+    });
+
+    assert(g.precomputed != null, 'usePrecomputed sets g.precomputed');
+  });
+
+  it('precomputed curves have been calculated correctly', function() {
```
Comment (unidwell) [inline test/use-precomputed-test.js:39]: This checks that `precomputed JSON` matched `computed.toJSON`. We could add the fromJSON test or test that `computed` matches `fromJSON(precomputed)`.

**Comment by mvayngrib** [comment]
@unidwell wow, that's some catch! Amazing that all the tests passed despite that. ~Unfortunately, that `pointFromJSON` will cost us all the perf gains (i'll double check).~

**Comment by mvayngrib** [comment]
@unidwell looks like there are still plenty of gains to be made (phew)

```js
// in RN, tested separately
elliptic.curves.p256.curve.pointFromJSON(precomputed)  // 300ms
elliptic.ec('p256')                                    // 995ms
```

another alternative to changing the ctor is to change use-precomputed from this:

```js
names.forEach(function(name) {
  if (!curves[name].g.precomputed) {
    curves[name].g.precomputed = precomputed[name];
  }
});
```


to this:

```js
names.forEach(function(name) {
  if (!curves[name].g.precomputed) {
    curves[name].g = curves[name].curve.pointFromJSON(precomputed[name]);
  }
});
```

edit: the benefit here is that the expensive `pointFromJSON` isn't run on every new instance of a curve

thoughts?

**Comment by unidwell** [comment]
```
names.forEach(function(name) {
  if (!curves[name].g.precomputed) {
    curves[name].g = curves[name].curve.pointFromJSON(precomputed[name]);
  }
});
```
> edit: the benefit here is that the expensive pointFromJSON isn't run on every new instance of a curve

This is cool :-) Tested it with changed `usePrecomputed` and correct precomputed data gets used.

**Comment by unidwell** [comment]
> // in RN, tested separately
elliptic.curves.p256.curve.pointFromJSON(precomputed)  // 300ms
elliptic.ec('p256')                                    // 995ms

Curious about the numbers. Is elliptic loaded lazily, contextually (within functions, selectors), or in global namespace? How many times is it called during app startup? Is 995ms per one core (process) and therefore multiple usages throttles other cores? Think mobile has one process if I remember correctly.

Is it only in the libs we control (we could change to lazy load global object) or is the majority in the libs outside our control? BNB libs, for example, can be done with lazy loading.

Lazy loading a singleton could improve the above gains. To do this, we could either provide a small library that we would use for our libs or do some monkey wrestling and change elliptic.ec to compute each object only once. This could be done outside elliptic.ec to reduce code inheritance/maintenance debt.

**Comment by mvayngrib** [comment]
>  Is elliptic loaded lazily, contextually (within functions, selectors), or in global namespace?

global to the neo-lib keypair module: https://github.com/ExodusMovement/neo/blob/master/packages/neo-lib/src/keypair.js#L4

we can load it lazily, like we did with ed25519 here: https://github.com/ExodusMovement/ripple/pull/27

> How many times is it called during app startup? 

once so far, for p256, the most expensive one. Multiple for some others. secp256k1 is cheap because it's precomputed by default. ed25519 is somewhat expensive (250ms), but i haven't figured out how to make it more efficient with precomputed 

> Is 995ms per one core 

mobile uses a single JS thread, so it blocks all other JS

> Is it only in the libs we control (we could change to lazy load global object) or is the majority in the libs outside our control? BNB libs, for example, can be done with lazy loading.

we've done some work deduplicating elliptic across node_modules/ on mobile, so hopefully soon it'll be one copy of elliptic (though not explicitly global). 

> Lazy loading a singleton could improve the above gains. To do this, we could either provide a small library that we would use for our libs or do some monkey wrestling and change elliptic.ec to compute each object only once. This could be done outside elliptic.ec to reduce code inheritance/maintenance debt.

a 2nd instance of a given curve from a single copy of elliptic doesn't cause the same perf drain (because `.precomputed` is stored on an object shared across instances). A separate library for monkeypatching precomputed curves was [discussed](https://github.com/ExodusMovement/elliptic/pull/1#discussion_r308926874). I was thinking that `usePrecomputed` would be called somewhere at the mobile/desktop root, so the modules below don't need to think too hard

**Comment by mvayngrib** [comment]
looks like there was a bug for `Point.prototype.toJSON` for short curves (thanks @unidwell for helping me look into that). I pushed a breaking test, fix, and recomputed curves, but i can separate that out into a separate PR, as this is getting a bit unwieldy, what say you?

**Comment by unidwell** [comment]
> I pushed a breaking test, fix, and recomputed curves, but i can separate that out into a separate PR, as this is getting a bit unwieldy, what say you?

Looks fine to me, no need to split into separate PR.

**Comment by mvayngrib** [comment]
for ref, upstream PR: https://github.com/indutny/elliptic/pull/188 (@fanatid graciously agreed to dig into the need for the change)

**Review by unidwell — APPROVED** [review]
tACK - nice toJSON improvement!

**Review by kklash — APPROVED** [review]
utACK

**Comment by mvayngrib** [comment]
probably going to rebase this on top of 6.4.1 (the new version all the cool kids are talking about now)

**Comment by mvayngrib** [comment]
rebased on top of 6.4.1

**Comment by ChALkeR** [comment]
~7cffa3b9e84ad2f32e9dd49426d20fa2c55b8122 was ignored with the rebase?~
~Do we need those files in the package?~
Upd: ah, only `lib/` is packaged, sorry.

Another question -- `hmac-drbg` is moved to a separate module upstream, is that expected?
Just to make sure that this doesn't affect this PR.

File: test/use-precomputed-test.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    var dir = path.join(__dirname, '..', 'lib/elliptic/precomputed')
+
+    function exportPoint(p) {
+      return JSON.parse(JSON.stringify(p.toJSON()));
+    }
+
+    fs.readdirSync(dir)
+      .filter(function (file) {
+        return path.extname(file) === '.json'
```
Comment (ChALkeR) [inline test/use-precomputed-test.js:50]: Why is `secp256k1` excluded from this test? It's specifically in `.js` format, not `.json`.

**Review by ChALkeR — DISMISSED** [review]
`secp256k1` precomputed curve should also be compared with computed during tests.
It looks reproducible with a simple change like `if (name === 'secp256k1') computed = computed[2]`, but that's probably not a good way there (see below).

Currently, modifying `secp256k1` precomputed curve file in a certain way still keeps the tests passing, that should not be the case.

---

Also: why is it split into a different `*.js` file instead of a json one?

Just to filter it out from e.g. tests based on the filename? Perhaps explicitly giving it a different suffix or a separate directory would be a better way to indicate that it uses a different format from all the other ones, than filtering based on `.js`/`.json` filename difference.

**Review by ChALkeR — APPROVED** [review]
Ok, I dismissed my prior review because, as @mvayngrib pointed out, `secp256k1` is precomputed upstream and not introduced here. Thanks for explanation!

I do not have any specific concerns with the code anymore, ~but I would still recommend adding a test for `secp256k1` precomputation~ upd: nevermind about that.

**Comment by mvayngrib** [comment]
@feri42 @roccomuso @unidwell do you guys want to take another look, or can we merge this?

**Comment by unidwell** [comment]
@mvayngrib Checked also changes since rebase, looks good.

**Review by sonaye — APPROVED** [review]
